fe/parameterize_batch.py

```diff
@@ -0,0 +1,523 @@
+
+                combined_pdb = Chem.CombineMols(Chem.MolFromPDBFile(host_pdb_file, removeHs=False), init_mol)
+                combined_pdb_str = StringIO(Chem.MolToPDBBlock(combined_pdb))
+                out_file = os.path.join(args.out_dir, "epoch_"+str(epoch)+"_insertion_deletion_"+host_name+"_conf_"+str(conf_idx)+".pdb")
+                writer = PDBWriter(combined_pdb_str, out_file)
+
+                v0 = np.zeros_like(x0)
+
+                parent_conn, child_conn = Pipe()
+                parent_conns.append(parent_conn)
+                # writer can be None if we don't care about vis
+                all_args.append([x0, v0, conf_idx % num_gpus, writer, child_conn])
+
+            processes = []
+
+            for arg in all_args:
+                p = Process(target=sim.run_forward_and_backward, args=arg)
+                p.daemon = True
+                processes.append(p)
+                p.start()
+
+            all_du_dls = []
+            for pc in parent_conns:
+                du_dls = pc.recv()
+                all_du_dls.append(du_dls)
+
+            all_du_dls = np.array(all_du_dls)
+            loss_grad_fn = jax.grad(error_fn, argnums=(0,)) 
+
+            error = error_fn(all_du_dls, T, lambda_schedule, true_dG)
+
+            print("---EPOCH", epoch, "---Train---LOSS", error)
+            with open(args.out_dir+"/loss.txt", "a+") as f:
+                f.write("---EPOCH %d----Train---LOSS---- %f\n\n " % (epoch, error))
+
+            error_grad = loss_grad_fn(all_du_dls, T, lambda_schedule, true_dG)
+            all_du_dl_adjoints = error_grad[0]
+
+            # send everything at once
+            for pc, du_dl_adjoints in zip(parent_conns, all_du_dl_adjoints):
+                pc.send(du_dl_adjoints)
+
+            # receive everything at once
+            all_dl_dps = []
+            for pc in parent_conns:
+                dl_dp = pc.recv()
+                all_dl_dps.append(dl_dp)
+
+            # terminate all the processes
+            for p in processes:
+                p.join()
+
+            all_dl_dps = np.array(all_dl_dps)
+            all_dl_dps = np.sum(all_dl_dps, axis=0)
+
+            allowed_groups = {vary_allowed_groups[epoch][0]: vary_allowed_groups[epoch][1]}
+
+            filtered_grad = []
+            for g_idx, (g, gp) in enumerate(zip(all_dl_dps, sim.system.param_groups)):
+                if gp in allowed_groups:
+                    pf = allowed_groups[gp]
+                    filtered_grad.append(g*pf)
+                    if g != 0:
+                        print("derivs", g_idx, '\t group', gp, '\t', g, '\t adjusted to', g*pf, '\t old val', sim.system.params[g_idx])
+                else:
+                    filtered_grad.append(0)
+
+            filtered_grad = np.array(filtered_grad)
+            opt_state = opt_update(epoch, filtered_grad, opt_state)
+
+        for data in test.iterbatches(1):
+
+            guest_mol = copy.deepcopy(data[0][0])
```

Comment:
The test set has a *lot* of duplicate code shared with the training set - can you find a way to refactor them out? If we change something in the training code we'd also have to remember to change it exactly the same way in the test code. 

For example
- the host_pdb code is identical, and can be moved entirely out of the training loop
- lambda schedules etc are the same.